tutorial/Execution.md

```diff
@@ -0,0 +1,412 @@
+
+At this point, we have most of the methods necessary to do some pretty useful work:
+
+```scala
+val config: Config = // ...
+val mode: Mode = // ...
+val loggingService: String => Unit = // ...
+
+val announceStart = Execution.from(loggingService("Starting job"))
+val announceStop = Execution.from(loggingService("Stopping job"))
+val scaldingJob = TypedPipe
+  .from(/* ... */)
+  .map(/* ... */)
+  .writeExecution(/* ... */)
+
+val scaldingJobWithAnnouncements =
+  announceStart.flatMap { _ =>
+    scaldingJob.flatMap { _ =>
+      announceStop
+    }
+  }
+
+scaldingJobWithAnnouncements.run(config, mode)
+```
+
+As always, remember that we are creating a plan to run a Scalding job (and logging service announcements). not actually running this code. Until we've called `Execution`'s `run` method, no work has been done (besides instantiating the loggingService. Also note that while code run in the `TypedPipe` methods `map`, `filter`, etc. may happen in a different run environment according on the `Mode` (e.g. remote Hadoop cluster), everything else is happening locally such as the calls to the `loggingService`.
+
+## The Real Execution
+
+Thus far, we have only discussed the simplified version of `Execution` to make it easier to understand. Here, we will talk about the real `Execution` in Scalding and how to make best use of it.
+
+### Executions are actually asynchronous
+
+Our `case class Execution` above does all work synchronously--only one thing at a time. In reality, `Execution` uses scala's `scala.concurrent.ExecutionContext` to schedule tasks asynchronously. So, if we want to turn non-Scalding asynchronous tasks into planned `Execution`s for use with Scalding, we can use the `Execution` companion object's method `def fromFuture[T](fn: ExecutionContext => Future[T]): Execution[T]` to schedule a scala `Future` on the same `ExecutionContext` that the `Execution` is using:
+
+```scala
+Execution.fromFuture { implicit executor: ExecutionContext =>
+  Future { /* ... asynchronous task ... */ }
+}
+```
+
+Additionally, the `run` method from our simplified `Execution` above, in the real `Execution`, takes in a scala `ExecutionContext` and yields a `Future`, so we can actually run `Execution` alongside other asynchronous scala code:
+
+```scala
+val config: Config = // ...
+val mode: Mode = // ...
+
+val scaldingJob: Execution[Unit] = // ...
+implicit val executor: ExecutionContext = // ...
+
+Future { /* ... asynchronous task ... */ }
+  .flatMap{ _ => scaldingJob.run(config, mode)(executor) }
+```
+
+But in general, we recommend that one extends `ExecutionApp` (below) and plans asynchronous actions from within that environment instead of integrating `Execution`s into existing asynchronous environments.
```

Comment:
@johnynek, what are your feelings on this? We use `ExecutionApp` pretty much exclusively at Twitter, so I don't want to misrepresent best practices across the community. E.g. are there teams that make use of services like `A => Future[(Config, Mode)]` to schedule `Execution`s alongside other async code?